src/lib.rs

```diff
@@ -710,26 +765,56 @@ impl SystemHandle {
     /// Subscribe given `recipient` to events of type `E`. See [`Context::subscribe()`].
     pub fn subscribe_recipient<M: 'static, E: Event + Into<M>>(&self, recipient: Recipient<M>) {
         let mut event_subscribers = self.event_subscribers.write();
+        event_subscribers.subscribe_recipient::<M, E>(recipient);
+    }
 
-        let subs = event_subscribers.events.entry(TypeId::of::<E>()).or_default();
+    /// Subscribe given `recipient` to events of type `E` and send the last cached event to it.
+    /// See [`Context::subscribe_and_receive_latest()`].
+    pub fn subscribe_and_receive_latest<M: 'static, E: CacheableEvent + Into<M>>(
+        &self,
+        recipient: Recipient<M>,
+    ) -> Result<(), SendError> {
+        let mut event_subscribers = self.event_subscribers.write();
 
-        subs.push(Box::new(move |e| {
-            if let Some(event) = e.downcast_ref::<E>() {
-                let msg = event.clone();
-                recipient.send(msg.into())?;
+        // Send the last cached value if there is one.
+        if let Some(last_cached_value) = event_subscribers.last_value_cache.get(&TypeId::of::<E>())
+        {
+            if let Some(msg) = last_cached_value.downcast_ref::<E>() {
+                recipient.send(msg.clone().into())?;
             }
+        }
 
-            Ok(())
-        }));
+        event_subscribers.subscribe_recipient::<M, E>(recipient);
+        Ok(())
     }
 
     /// Publish an event. All actors that have previously subscribed to the type will receive it.
     ///
+    /// The event will be cached if it implements [`CacheableEvent`]. Actors that will subscribe to
+    /// such an event type in future may choose to receive the last cached event upon subscription.
+    ///
     /// When sending to some subscriber fails, others are still tried and vec of errors is returned.
     /// For direct, non-[`Clone`] or high-throughput messages please use [`Addr`] or [`Recipient`].
     pub fn publish<E: Event>(&self, event: E) -> Result<(), PublishError> {
-        let event_subscribers = self.event_subscribers.read();
-        if let Some(subs) = event_subscribers.events.get(&TypeId::of::<E>()) {
+        let type_id = TypeId::of::<E>();
+
+        let event_subscribers = if let Some(cached) = event.cache() {
+            // This write lock is a possible contention point if many actors publish events at high
+            // rate at the same time. If that ever becomes a bottleneck, we could transition to
+            // a hash map with finer locking, like dashmap. Like in anx; intermediate version of #87.
```

Comment:
> Like in anx

what is anx?

> intermediate version of #87.

We should just link the particular commit.

Reply:
anx is a typo, will fix. I plan to change the other part anyway..